src/drivers/rotor_frequency/pcf8583/pcf8583.cpp

```diff
@@ -0,0 +1,702 @@
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file pcf8583.cpp
+ *
+ * @author Vít Hanousek <slimonslimon@gmail.com>
+ *
+ * Driver for Main Rotor speed sensor using MLAB.CZ RTC003A I2C counter.
+ */
+
+#include <px4_platform_common/px4_config.h>
+#include <px4_platform_common/defines.h>
+#include <px4_platform_common/getopt.h>
+#include <px4_platform_common/workqueue.h>
+
+#include <lib/parameters/param.h>
+
+#include <board_config.h>
+
+#include <px4_log.h>
+
+#include <drivers/device/i2c.h>
+
+
+#include <sys/types.h>
+/*#include <stdint.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <semaphore.h>
+
+#include <fcntl.h>
+#include <poll.h>
+#include <stdio.h>
+#include <math.h>
+#include <unistd.h>
+#include <vector>
+
+
+
+#include <perf/perf_counter.h>
+*/
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+
+#include <uORB/uORB.h>
+#include <uORB/topics/rotor_frequency.h>
+
+#include <drivers/drv_hrt.h>
+#include <drivers/drv_sensor.h>
+
+
+/* Configuration Constants */
+#define PCF8583_BUS_DEFAULT 		      PX4_I2C_BUS_EXPANSION
+#define PCF8583_DEVICE_PATH	         "/dev/pcf8583"
+
+
+//loaded parameters defaluts
+#define PCF8583_BASEADDR_DEFAULT 	            0x51     //ROTROFREQ_ADDR param default value
+#define PCF8583_POOL_INTERVAL_DAFAULT	         1000000  //ROTORFREQ_POOL param defalut value
+#define PCF8583_MAGNET_COUNT_DEFAULT           1        //ROTORFREQ_MAGNET param defalut value
+#define PCF8583_RESET_COUNT_DEFAULT            0 //0 - reset after every measurement
+
+#define PCF8583_POOL_INTERVAL_MAX              10000000 //10s max - limiter for setting by ioctl
+
+
+#ifndef CONFIG_SCHED_WORKQUEUE
+# error This requires CONFIG_SCHED_WORKQUEUE.
+#endif
+
+
+class PCF8583 : public device::I2C
+{
+public:
+	PCF8583(int bus = PCF8583_BUS_DEFAULT,
+	      int address = PCF8583_BASEADDR_DEFAULT);
+	virtual ~PCF8583();
+
+	virtual int 		init();
+
+	virtual ssize_t		read(device::file_t *filp, char *buffer, size_t buflen);
+	virtual int			   ioctl(device::file_t *filp, int cmd, unsigned long arg);
+
+	/**
+	* Diagnostics - print some basic information about the driver.
+	*/
+	void				print_info();
+
+protected:
+	virtual int			probe();
+
+private:
+	int            _pool_interval; //Interval of reading counter and publishing frequency
+   int            _pool_interval_default; //Interval of reading counter and publishing frequency
+   float          _indicated_frequency;
+   float          _estimated_accurancy;
+   int            _count;
+   int            _reset_count;
+   int            _magnet_count;
+   uint64_t       _lastmeasurement_time;
+	work_s			_work{};
```

Comment:
Please use the work queue within PX4. Here's an example - https://github.com/PX4/Firmware/blob/47cab3dba52520c4c3e9b42529152baf1148f1ee/src/drivers/lights/rgbled/rgbled.cpp#L63

Reply:
Done.